File: May-LeetCoding-Challange-2020/Week3/328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # Sol 1. Linked List
    def oddEvenList(self, head: ListNode) -> ListNode:
        # Check if the current node is valid or not
        # Base case
        if(not head):
            return head
        
        odd = head
        even = odd.next
        evenList = even
        
        while(even and even.next):
            odd.next = even.next
            odd = odd.next
            
            even.next = odd.next
            even = even.next
            
        odd.next = evenList
        return head
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    # Run Code Runtime: 48 ms
    # Submit Runtime: 68 ms
    # Submit Memory: 15.7 MB
    # Submit Runtime Distribution: 7.57% faster, fuck this 

File: May-LeetCoding-Challange-2020/Week3/328-Odd_Even_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:


    /* Time Complexity: O(n)
     * Space Complexity: O(1)
     * Runtime: 20ms 58.38%
     * Memory Usage: 9.8 MB 
    */
    ListNode* oddEvenList(ListNode* head) {
        
        if(head == NULL || head->next == NULL){
            return head;
        }
        
        ListNode *root = head;
        ListNode *last_node = NULL;
        ListNode *temp = NULL;
        int count = 1;
        
        while(head->next != NULL){
            count++;
            head = head->next;
        }
        
        last_node = head;
        head = root;
        temp = head->next;
        
        for(int i = 1; i < count; i+=2){

            if(temp == last_node)
                break;
            
            if(i % 2 == 1){
                head->next = temp->next;
                temp->next = NULL;
                last_node->next = temp;
                last_node = temp;
            }
            
            
            head = head->next;
            temp = head->next;
            
        }
        
        return root;
    }
};